pkg/analysis/completion.go

```diff
@@ -277,6 +285,123 @@ func (a *Analyzer) keywordArgSymbols(fn query.Signature, args callArguments) []q
 	return symbols
 }
 
+// Find the object part of an attribute expression that has a dot '.' immediately before the given point.
+func (a *Analyzer) findAttrObjectExpression(nodes []*sitter.Node, pt sitter.Point) *sitter.Node {
+	if pt.Column == 0 {
+		return nil
+	}
+
+	var dot *sitter.Node
+	searchRange := sitter.Range{StartPoint: sitter.Point{Row: pt.Row, Column: pt.Column - 1}, EndPoint: pt}
+	var parentNode *sitter.Node
+	for i := len(nodes) - 1; i >= 0; i-- {
+		parentNode = nodes[i]
+		dot = query.FindChildNode(parentNode, func(n *sitter.Node) int {
+			if query.PointBeforeOrEqual(n.EndPoint(), searchRange.StartPoint) {
+				return -1
+			}
+			if n.StartPoint() == searchRange.StartPoint &&
+				n.EndPoint() == searchRange.EndPoint &&
+				n.Type() == "." {
+				return 0
+			}
+			if query.PointBeforeOrEqual(n.StartPoint(), searchRange.StartPoint) &&
+				query.PointAfterOrEqual(n.EndPoint(), searchRange.EndPoint) {
+				return 0
+			}
+			return 1
+		})
+		if dot != nil {
+			break
+		}
+	}
+	if dot != nil {
+		expr := parentNode.PrevSibling()
+		for n := dot; n != parentNode; n = n.Parent() {
+			if n.PrevSibling() != nil {
+				expr = n.PrevSibling()
+				break
+			}
+		}
+
+		if expr != nil {
+			a.logger.Debug("dot completion",
+				zap.String("dot", dot.String()),
+				zap.String("expr", expr.String()))
+			return expr
+		}
+	}
+	return nil
+}
+
+// Perform some rudimentary type analysis to determine the Starlark type of the node
+func (a *Analyzer) analyzeType(doc document.Document, node *sitter.Node) string {
+	switch node.Type() {
+	case query.NodeTypeString:
+		return "String"
+	case query.NodeTypeDictionary:
+		return "Dict"
+	case query.NodeTypeList:
+		return "List"
+	case query.NodeTypeIdentifier:
+		sym, found := a.FindDefinition(doc, node, doc.Content(node))
+		if found {
+			switch sym.Kind {
+			case protocol.SymbolKindString:
+				return "String"
+			case protocol.SymbolKindObject:
+				return "Dict"
+			case protocol.SymbolKindArray:
+				return "List"
+			}
+		}
+	case query.NodeTypeCall:
+		fnName := doc.Content(node.ChildByFieldName("function"))
+		sig, found := a.signatureInformation(doc, node, fnName)
+		if found && sig.ReturnType != "" {
+			switch strings.ToLower(sig.ReturnType) {
+			case "str", "string":
+				return "String"
+			case "list":
+				return "List"
+			case "dict":
+				return "Dict"
+			default:
+				return sig.ReturnType
+			}
+		}
+	}
+	return ""
+}
+
+func (a *Analyzer) availableMembers(doc document.Document, node *sitter.Node) []query.Symbol {
+	if t := a.analyzeType(doc, node); t != "" {
+		if class, found := a.builtins.Types[t]; found {
+			return class.Members
+		}
+	}
+	return a.builtins.Members
```

Comment:
Yeah, it's worth some more thought. I felt like erring on the side of too much info is more helpful even if some of the options are incorrect but if that's more likely to send people in the wrong direction then maybe it's better to be more conservative? If we think about combining this with error messages from a running Tilt, would that be odd to be given an incorrect completion item and then have Tilt report an error because the method wasn't there?

Maybe there's some way to indicate in the completion list that we're offering this as an option but it might not be accurate for the actual runtime type.